### DISCUSSION TITLE
mapview staying blank (except scale bar)

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/vtm#908] [Q&A] [answered]
I'm trying to integrate VTM into an existing Android app. I started with the `MapsforgeActivity` example, and in the beginning it seemed to work (using a map file stored in the emulated device). Then I did a couple of generalizations (as I have to support other map sources besides VTM), and while _they_ do work now, I somehow broke VTM: The map view is completely white except the scale bar being displayed bottom-left. Though the code still looks pretty close to the original example. :-(
Logcat does not provide specific error messages related to VTM.

Question is: Is there a way to debug for what's going wrong here, what's missing etc.? Any object where I can read the "health state" of VTM from, or something like this? Any pointer to get me looking in the right place would be very welcomed. Thanks in advance!

**Comment by devemux86** [comment, marked as answer]
It is better to study the [Samples](https://github.com/mapsforge/vtm/tree/master/vtm-android-example/src/org/oscim/android/test) and compare the code.
And implement map life cycle as seen in [GettingStarted](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/GettingStarted.java).

---

Are you going to use VTM in c:geo?

**Reply by moving-bits** [reply]
Have it back working, thanks. Culprit was a `zoomToBounds()` introduced by me, which got called with lat/lon swapped, and with coords slightly outside of Berlin (which I was using as a test map) ☹️.

> Are you going to use VTM in c:geo?

That would be the goal ultimately, but for now I just started playing around with it a bit to get used to it. We are currently supporting both Google Maps and the "old" Mapsforge map, both having their own implementations due to the history of c:geo, which is very hard to maintain and to keep in sync in terms of functionality and behavior. Also, Mapsforge implementation is relatively slow (compared to VTM), and we have found no way to make it rotating. From the first bit of playing with it VTM is way faster, supports additional map sources (which may come handy in the future) etc.

Anything we should keep in mind before starting to integrate it into c:geo?

**Comment by devemux86** [comment]
> We are currently supporting both Google Maps and the "old" Mapsforge map

2 different OpenGL map engines together?
This would be an interesting experiment.

> Anything we should keep in mind before starting to integrate it into c:geo?

The API is expected to be different in several functions.
e.g. markers use ItemizedLayer(s) (see [MarkerOverlayActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/MarkerOverlayActivity.java))

_I keep VTM compatible with Mapsforge maps / themes._

VTM themes are more advanced with extra rules.
Mapsforge themes may need minor adjustments to render better in OpenGL.
(can also be done with [code](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/MapsforgeActivity.java#L303-L322))

**Comment by moving-bits** [comment]
Ok, I've made some progress: Mapsforge offline maps do work, as does Google Maps. Switching between OSM / GM as well as between different offline maps does work as well. But I'm facing two issues currently:

### Online maps (bitmap map source using a URL)
Displaying an online map source does not work yet, the mapview is a dark gray area only (but with map scale). I'm using the following code fragment to initialize the map (copied from some of the examples):

```
public void addTileLayer(final Map map) {
    final OkHttpClient.Builder httpBuilder = new OkHttpClient.Builder();
    final Cache cache = new Cache(new File(LocalStorage.getExternalPrivateCgeoDirectory(), "tiles"), 20 * 1024 * 1024);
    httpBuilder.cache(cache);
    final TileSource tileSource = BitmapTileSource.builder()
        .url(mapUri.toString())
        .tilePath(tilePath)
        .zoomMax(zoomMax)
        .zoomMin(zoomMin)
        .httpFactory(new OkHttpEngine.OkHttpFactory(httpBuilder))
        .build();
    tileLayer = new BitmapTileLayer(map, tileSource);
    map.layers().add(tileLayer);
}
```
Anything I'm missing here? Something else I need to initialize etc.?

Second question is somewhat related:

### Switching between different tile providers

What do I have to do to switch between different tile providers? Something like `map.layers().clear()` seems to be a bit too radical (leads to crashes), but what would be the right way for cleaning up one tile source and leaving the map view in a state where it happily accepts a different tile source?

Thanks for your support!

**Comment by devemux86** [comment]
Please study the working samples, like [BitmapTileActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/BitmapTileActivity.java).
See how to construct bitmap tile sources in [DefaultSources](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/tiling/source/bitmap/DefaultSources.java).

---

Maps are layers, replace the existing with new ones (see also [here](https://github.com/mapsforge/vtm/blob/master/vtm-gdx/src/org/oscim/gdx/InputHandler.java#L148-L160)).
(like in Mapsforge)

**Reply by moving-bits** [reply]
Ok, thanks again. Exchanging
- `.httpFactory(new OkHttpEngine.OkHttpFactory(httpBuilder))` 
  (which was actually taken from `vtm-android-example/.../BaseMapActivity.java`)

with 
- `tileSource.setHttpEngine(new OkHttpEngine.OkHttpFactory(httpBuilder));`
  (as in `vtm-android-example/.../BitmapTileActivity.java`)

did the trick.

Will look into handling tile layers over the next days. I guess I'll have to make every tilesource remember which layers they have set, so that on swapping tilesources they can suspend and remove their layers first, before other tilesources can add theirs.

**Reply by moving-bits** [reply]
> Will look into handling tile layers over the next days. I guess I'll have to make every tilesource remember which layers they have set, so that on swapping tilesources they can suspend and remove their layers first, before other tilesources can add theirs.

FYI: Changing tile sources does work now across all map sources (VTM bitmap OSM online, VTM vector `.map` offline, Google Maps). Beside remembering and removing VTM layers and base map I needed to call `clearMap` to refresh the display.
